index.ts

```diff
@@ -1,213 +1,170 @@
-
-  return file_metadata_event && (
-    is_publisher(file_metadata_event, auth_event) ||
-    is_delegator(file_metadata_event, auth_event)
-  );
-};
-
-const get_put_permission = async (
-  env: WorkerEnv,
-  auth_event: Event,
-): Promise<boolean> => {
-  const pubkey = auth_event.pubkey;
-  const role: Role | undefined = await env.BANBOORU_PUBKEY_ROLE_KV.get(
-    pubkey,
-  ) as Role;
-  return ["admin", "user"].includes(role);
-};
-
-export default {
-  async fetch(request: Request, env: WorkerEnv, _ctx: any): Promise<Response> {
-    const url = new URL(request.url);
-
-    if (!url.pathname.startsWith("/file/")) {
-      return new Response("Not Found.", { status: 404 }); // 404 Not Found
+}
+
+/** A  WorkerEnv once the request as been authenticated (using NIP-98) */
+type AuthEnv = WorkerEnv & {
+  /** Verified `kind 27235` event from the request Authorization header. */
+  authEvent: AuthEvent
+  /** The user id, same as `authEvent.pubkey` */
+  userId: string
+  /** The user's role (as retrieved with `BANBOORU_PUBKEY_ROLE_KV.get(userId)`. */
+  userRole: Promise<Role | undefined>
+}
+
+/**
+ * Authentication middleware. Requires a NIP-94 Authorization header. 
+ */
+function auth(next: MatchHandler<AuthEnv>): MatchHandler<WorkerEnv> {
+  return async (request: Request, env: HandlerContext<WorkerEnv>, params: Record<string, string>) => {
+    const authEvent: AuthEvent | { error: string } = await get_auth_event(request);
+    if (Object.hasOwn(authEvent, "error")) {
+      // TODO: Return JSON body with error message.
+      return new Response(`Unauthorized: ${authEvent.error}`, { status: 401 });
+    } else {
+      const userId = authEvent.pubkey
+      const userRole = env.BANBOORU_PUBKEY_ROLE_KV.get(authEvent.pubkey) as Promise<Role | undefined>
+      return next(request, { ...env, authEvent, userId, userRole }, params)
     }
-
-    const path_file_sha256_hex = url.pathname.slice(6);
-    if (!path_file_sha256_hex.match(/^[A-Fa-f0-9]{64}$/)) {
-      return new Response("/file/<FILE-SHA-256> could not validate SHA-256", {
-        status: 400,
-      }); // 400 Bad Request
+  }
+}
+
+/**
+ * Restricted access middleware. Requires a NIP-94 Authorization header from a known pubblic key with one of the specific `roles`.
+ */
+function restricted(roles: Role[], next: MatchHandler<AuthEnv>): MatchHandler<WorkerEnv> {
+  return auth(async (request: Request, env: HandlerContext<AuthEnv>, params: Record<string, string>) => {
+    const role = await env.userRole
+    if(role && roles.includes(role)) {
+      return next(request, env, params)
+    } else {
+      // TODO: Return JSON body with error message.
+      return new Response("Forbidden.", { status: 403 });
     }
+  })
+}
 
-    // GET /file/<SHA-256>
-    if (request.method === "GET") {
-      const object = await env.BANBOORU_BUCKET.get(path_file_sha256_hex);
+export default {
+  fetch: router<WorkerEnv>({
+    'GET /file/:hash': async function getFile(_req, env, params) {
```

Comment:
just a minor detail but these functions could not be named and be shortened to:

```
    'GET /file/:hash': async (_req, env, params) => {
```

Reply:
I prefer to name my function, so the name can appear in the stack trace when logging errors.